### PR TITLE
Embed Block: responsive design improvements

### LINKED ISSUE #3581: Embed Block Responsiveness
## Feature Description

We received [a report on Twitter](https://twitter.com/KennyLeys/status/1288795810818396166) that the embed block wasn't fully responsive.

WordPress and the WordPress block editor handle responsive embeds in various way.

First of all, they [require themes to opt in](https://developer.wordpress.org/block-editor/developers/themes/theme-support/#responsive-embedded-content). But even without that, there is _some_ responsiveness for embeds.

Then there are differences between iframes like the WordPress embeds and embeds with a fixed aspect ratio like YouTube videos.

They do it [with this ol' trick](https://css-tricks.com/aspect-ratio-boxes/) using padding and pseudo elements.

Some code references:

https://github.com/WordPress/gutenberg/blob/4fe937c2188e0677b03e6eb5d2c54292705f6c45/packages/block-library/src/embed/style.scss#L37-L84
https://github.com/WordPress/gutenberg/blob/4857ad58c1241b3d63d21a6880c989b85746c3dc/packages/block-library/src/embed/util.js#L132-L185

Our embed block has one big difference to that: users can set width & height to their liking, so we have to do the aspect ratio calculation on the fly.

This is not really an issue with the embed block itself, but more of a question about how to style the player correctly.

**Possible solution**

1. Add a new `div` with the class  name `wp-block-embed__wrapper` around `<amp-story-player>`
1. Give the player itself the following inline style:  
    ```css
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    ```
1. Give the `wp-block-embed__wrapper` element the following inline style:   
    ```css
    position: relative;
    overflow: hidden;
    padding-top: calc(600 / 360 * 100%);
    height: 0;
    ``` 
    where 600 and 360 are the user provided numbers, hence the `calc()`

This is basically the Gutenberg solution but with support for arbitrary aspect ratios. That's why we cannot use a pseudo element here, since pseudo elements cannot be styled via inline styling.

[Using custom properties](https://css-tricks.com/aspect-ratio-boxes/#using-custom-properties) for the aspect ratio could be an alternative.

## Alternatives Considered

a) Don't let users edit the height, maintain a fixed aspect ratio
b) Solve this in the `<amp-story-player>` component

## Additional Context

Player docs: https://github.com/ampproject/amphtml/blob/master/spec/amp-story-player.md

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Implementation Brief

* Add new wrapper elements to markup
* Add new external stylesheet for embed block styling -> the only inline style will be for `max-width` and setting `--aspect-ratio`
* Use the new `Embed` class in Admin.php to generate markup for embed block.

### PR DESCRIPTION
**Pull request by swissspidy**
## Summary

- Fix styling of blocks aligned in the center
- Fix styling of blocks aligned left or right
- Ensure blocks scale proportionally depending on available height
- For AMP, uses responsive layout to achieve the same

## Relevant Technical Choices

* CSS is added as inline style as we don't yet have CSS processing in place to have this as an external stylesheet. Plus, it saves an extra network request.

## To-do

**AMP**

* [x] Address AMP validation error for `<amp-story-player>` with `layout="responsive"`  
    Getting the following error:  
    ```
    Tag 'i-amphtml-sizer' is disallowed as child of tag 'amp-story-player'. Child tag must be one of ['a']
    ```
    Blocked by https://github.com/ampproject/amphtml/issues/30414
* [x] Address issue where `<amp-story-player>` has width and height hardcoded.  
    Blocked by https://github.com/ampproject/amphtml/issues/30423

## User-facing changes

Embed block is now properly responsive.

## Testing Instructions

Non-AMP:

1. Insert Web Stories embed block in post, make it quite large
1. View post on small screen sizes, verify that the block shrinks accordingly

For AMP:

1. Activate AMP plugin
1. Switch to nightly channel at https://cdn.ampproject.org/experiments.html
1. View AMP version of the post
1. Verify that the block shrinks accordingly
1. Verify that there are no AMP validation errors for the post.

---

Fixes #3581

### CODE REVIEW

**Comment by swissspidy** [comment]
@westonruter As mentioned in the PR description, I am getting AMP validation errors with `<amp-story-player>` and responsive layout, probably caused by the Optimizer. Is this a bug in the AMP plugin, the spec, or elsewhere?

@enriqe If your remember my question from #3581 last month... I've finally got around to implement this. `layout="responsive"` is what I need, but the `amp-story-player` CSS seems to force a width of 360px and height of 600px for the component.

So even if I use `<amp-story-player width="410" height="684" layout="responsive">...`, the width and height stay at 360/600.

It seems to be caused by this CSS that gets applied:

https://github.com/ampproject/amphtml/blob/8fb1a06a49088fb87de9b24646dee2058d910a87/css/amp-story-player.css#L17-L22

Is this a known issue?

**Comment by westonruter** [comment]
> As mentioned in the PR description, I am getting AMP validation errors with `<amp-story-player>` and responsive layout, probably caused by the Optimizer. Is this a bug in the AMP plugin, the spec, or elsewhere?

This is indeed appears to be a problem with the validator spec. I tried running the `amp-story-player` through the Node.js Optimizer and it produced the same error. I've filed an issue: https://github.com/ampproject/amphtml/issues/30414.

Another issue I noticed. In the AMP plugin's Transitional mode (when SSR is disabled):

```html
  <amp-story-player width="360" height="600" layout="responsive">
    <a href="https://wordpressdev.lndo.site/web-stories/story-of-bison/?amp" rel="amphtml" data-amp-original-style="--story-player-poster: url(https://wordpressdev.lndo.site/content/uploads/2020/07/American_bison_k5680-1-scaled.jpg)" class="amp-wp-a8a8a7a">
      Story of Bison
    </a>
  </amp-story-player>
```

When AMP-to-AMP linking is enabled, the `<amp-story-player>` link is pointing to `?amp` which is not correct.

I suggest that an `amp_to_amp_linking_element_excluded` filter to prevent linking to `?amp`:

```php
add_filter(
	'amp_to_amp_linking_element_excluded',
	function ( $excluded, $url, $rel, $element ) {
		if ( $element instanceof DOMElement && $element->parentNode && 'amp-story-player' === $element->parentNode->tagName ) {
			$excluded = true;
		}
		return $excluded;
	},
	10,
	4
);
```

Otherwise each request to populate the `amp-story-player` will result in a redirect from `./?amp` to `./`.

**Comment by swissspidy** [comment]
Thanks @westonruter!

Filtering `amp_to_amp_linking_element_excluded` sounds good. Also, I think I can work around `The property '--story-player-poster' in attribute 'style' in tag 'a' is disallowed.` with an inline style tag as https://github.com/ampproject/amphtml/issues/24262 seems quite stale.

**Comment by spacedmonkey** [comment]
@swissspidy Are we pushing for responsive now for this embed blocks? Should we remove the height / width options on these blocks then?

File: includes/Embed_Block.php
Comment (spacedmonkey) [inline includes/Embed_Block.php]: Why is this an inline style? Just wondering.

File: includes/Story_Renderer/Embed.php
Comment (spacedmonkey) [inline includes/Story_Renderer/Embed.php]: I don't like how this references another class and requires that class to be run for this to work. Don't you also need the method `init_assets` to run here. I think the contents of `init_assets` should be moved to trait so both classes can reference this repeated logic?

File: includes/Story_Renderer/Embed.php
Comment (spacedmonkey) [inline includes/Story_Renderer/Embed.php:79]: If we are changing styling like this, we will have to note this in the change log?

File: includes/Story_Renderer/Embed.php
Comment (spacedmonkey) [inline includes/Story_Renderer/Embed.php]: Is there an issue with large float numbers. Should be try and limit this to say 3/5 DP?

File: includes/Embed_Block.php
Comment (swissspidy) [reply includes/Embed_Block.php]: See "Relevant Technical Choices" in the pull request description

Comment (spacedmonkey) [reply includes/Embed_Block.php]: Can we add a `todo` or comment to explain this in the code?

**Comment by spacedmonkey** [comment]
Can we incorpate https://github.com/google/web-stories-wp/pull/3879 in this PR?

File: includes/Story_Renderer/Embed.php
Comment (swissspidy) [reply includes/Story_Renderer/Embed.php]: Default here is 6, which I think is fine.

Comment (spacedmonkey) [reply includes/Story_Renderer/Embed.php]: I am not sure I understand your answer here. What users can define the height and width. What if I define the height and width as. 609/137 resulting in 4.44525547445

File: includes/Story_Renderer/Embed.php
Comment (swissspidy) [reply includes/Story_Renderer/Embed.php]: `init_assets ` should not really be reusable / run twice though, so a trait isn't the best idea.

Scripts and styles only need to be registered **once**.

The `Embed` class is instantiated many times, but `Embed_Block` is only instantiated once in the `Plugin` class.

If anything, there could be a new `Assets` class or anything that registers all commonly used scripts/styles.

Then `Embed_Block` can enqueue just the scripts/styles it needs, and `Embed` enqueues the scripts/styles it needs.

File: includes/Story_Renderer/Embed.php
Comment (swissspidy) [reply includes/Story_Renderer/Embed.php:79]: This PR itself will be listed in the changelog (as most other PRs), but this markup change is not breaking and does not really warrant a special callout.

**Comment by swissspidy** [comment]
> Are we pushing for responsive now for this embed blocks? Should we remove the height / width options on these blocks then?

Width/height are still needed to set the aspect ratio and the maximum size of the block.

**Comment by swissspidy** [comment]
> Can we incorpate #3879 in this PR?

See https://github.com/google/web-stories-wp/pull/3879#issuecomment-696811011 for why this is not as easy as it sounds. It will easily cause block deprecations.

To give an example:

Let's say we use `Embed` in `Admin`.

Tomorrow we change `Embed.php` to render some additional buttons to make the embed render block more awesome. It will render markup that looks like this (simplified):

```
<div class="wp-block-web-stories-embed <?php echo esc_attr( $align ); ?>">
	<div class="wp-block-embed__wrapper" style="<?php echo esc_attr( $wrapper_style ); ?>">
		<amp-story-player>
			<a
				href="<?php echo esc_url( $url ); ?>"
				style="<?php echo esc_attr( $poster_style ); ?>">
				<?php echo esc_html( $title ); ?>
			</a>
		</amp-story-player>
	</div>
	<button>Reload Story</button>
	<button>View all stories</button>
	<button>Click me!</button>
</div>
```

`Admin` will end up printing something like this:

```
<!-- wp:web-stories/embed {"url":"%1\$s","title":"%2\$s","poster":"%3\$s"} -->
<div class="wp-block-web-stories-embed <?php echo esc_attr( $align ); ?>">
	<div class="wp-block-embed__wrapper" style="<?php echo esc_attr( $wrapper_style ); ?>">
		<amp-story-player>
			<a
				href="<?php echo esc_url( $url ); ?>"
				style="<?php echo esc_attr( $poster_style ); ?>">
				<?php echo esc_html( $title ); ?>
			</a>
		</amp-story-player>
	</div>
	<button>Reload Story</button>
	<button>View all stories</button>
	<button>Click me!</button>
</div>
<!-- /wp:web-stories/embed -->
```

Now, the block's `save()` function does not recognize this code. **Unless**, of course, the `save` function also prints exactly this code. That's what we should avoid though. It would mean that `Embed` is no longer an enhanced version of `save`, but just a copy of it.

**Comment by westonruter** [comment]
> Also, I think I can work around `The property '--story-player-poster' in attribute 'style' in tag 'a' is disallowed.` with an inline style tag as [ampproject/amphtml#24262](https://github.com/ampproject/amphtml/issues/24262) seems quite stale.

I don't think the CSS variable will actually be a problem for us because the AMP plugin automatically moves all `style` attributes to the `style[amp-custom]` anyway. So that aspect is primarily a problem for non-WP sites. 

The part that remains a problem for us is the `i-amphtml-sizer`.

File: includes/Story_Renderer/Embed.php
Comment (swissspidy) [reply includes/Story_Renderer/Embed.php]: Oh, I thought you were talking about the number of decimals for `%F`

This PR doesn't change existing behavior for the embed block's min/max widthheight, users can input whatever they like.

Some sane limits sound reasonable, but that'd be a new ticket.

**Comment by Enriqe** [comment]
> @westonruter As mentioned in the PR description, I am getting AMP validation errors with `<amp-story-player>` and responsive layout, probably caused by the Optimizer. Is this a bug in the AMP plugin, the spec, or elsewhere?
> 
> @Enriqe If your remember my question from #3581 last month... I've finally got around to implement this. `layout="responsive"` is what I need, but the `amp-story-player` CSS seems to force a width of 360px and height of 600px for the component.
> 
> So even if I use `<amp-story-player width="410" height="684" layout="responsive">...`, the width and height stay at 360/600.
> 
> It seems to be caused by this CSS that gets applied:
> 
> https://github.com/ampproject/amphtml/blob/8fb1a06a49088fb87de9b24646dee2058d910a87/css/amp-story-player.css#L17-L22
> 
> Is this a known issue?

Sorry, was OOO. Hadn't heard of this one before, I just filed https://github.com/ampproject/amphtml/issues/30423 and will be looking into this.

File: includes/Story_Renderer/Embed.php
Comment (spacedmonkey) [reply includes/Story_Renderer/Embed.php]: As mentioned in https://github.com/google/web-stories-wp/issues/3135 there is now a possiblity that there will be a third place where this script will be refenereced.

Comment (swissspidy) [reply includes/Story_Renderer/Embed.php]: Added a to-do note for this so we can split this out into its own class in a separate ticket & PR

File: includes/Embed_Block.php
Comment (spacedmonkey) [inline includes/Embed_Block.php]: @swissspidy Does this CSS snippet go through the css linter? Just wonder. If not, I believe it should. Maybe it should be write in a CSS file and loaded into memory?

**Comment by spacedmonkey** [comment]
I think CSS linter should should run on these changes.

File: includes/Embed_Block.php
Comment (swissspidy) [reply includes/Embed_Block.php]: I'd rather actually load this as an external stylesheet instead of doing `file_get_contents` or the like here.

Comment (swissspidy) [reply includes/Embed_Block.php]: @spacedmonkey thoughts?

Comment (spacedmonkey) [reply includes/Embed_Block.php]: Isn't there a performance reason you didn't do that in the first place. For the record, I don't love using `file_get_contents` here, but is this best for the user.

Comment (swissspidy) [reply includes/Embed_Block.php]: How is using file_get_contents better for the user?

I will change it, but just using stylelint is not much added value. Ideally we‘d minify CSS etc.

Comment (westonruter) [reply includes/Embed_Block.php]: See also how this is being done in the AMP plugin:

https://github.com/ampproject/amp-wp/blob/7a41a7b3bb06972ca73cfec8dbe693635cc0499d/src/MobileRedirection.php#L418-L419

It can pull in the built CSS file rather than the source one.

**Comment by swissspidy** [comment]
@spacedmonkey Moved it to a CSS file now 👍  Will now wait on the shortcode being merged so I can fix conflicts here.

**Comment by spacedmonkey** [comment]
This is looking good. I will await to approve it until shortcode is merged.

**Comment by swissspidy** [comment]
@spacedmonkey resolved merge conflicts and tested everything again, mind taking another look?

cc @westonruter for reviewing AMP-related changes in case you have any more input here

File: includes/Embed_Base.php
Comment (westonruter) [inline includes/Embed_Base.php]: Why the `if ( $css )`? Is that to handle the case where `embed.css` is not built? If so, wouldn't it be better to use a `file_exists()` check around this entire block?

File: includes/Story_Renderer/Embed.php
Comment (westonruter) [inline includes/Story_Renderer/Embed.php]:
```suggestion
		$poster = ! empty( $this->story->get_poster_portrait() ) ? esc_url_raw( $this->story->get_poster_portrait() ) : '';
```

Since `esc_url_raw()` is for sanitization. The escaping is done below when `$poster_style` is output inside of the `style` attribute.

File: includes/Story_Renderer/Embed.php
Comment (westonruter) [inline includes/Story_Renderer/Embed.php]:
```suggestion
			(int) $args['width'],
			(int) $args['height']
```

Comment (westonruter) [reply includes/Story_Renderer/Embed.php]: I suppose in reality this is overkill since `%d` string format is used.

File: includes/Story_Renderer/Embed.php
Comment (westonruter) [inline includes/Story_Renderer/Embed.php]:
```suggestion
			<div class="<?php echo esc_attr( "$class web-stories-embed web-stories-embed-amp $align" ); ?>">
```

File: includes/Story_Renderer/Embed.php
Comment (westonruter) [inline includes/Story_Renderer/Embed.php]:
```suggestion
		<div class="<?php echo esc_attr( "$class web-stories-embed $align" ); ?>">
```

File: includes/Embed_Base.php
Comment (swissspidy) [reply includes/Embed_Base.php]: There’s no build step, but `file_get_contents` can return false

Comment (westonruter) [reply includes/Embed_Base.php]: But it will also issue a PHP warning at the same time.

Comment (swissspidy) [reply includes/Embed_Base.php]: `file_exists` sounds reasonable, but without the `$css` check PHPStan will still rightfully say "wp_add_inline_style expects string, string|false given."

Comment (westonruter) [reply includes/Embed_Base.php]: Got it. Could also to `is_string( $css )` I suppose.

**Review by westonruter — APPROVED** [review]

File: includes/Embed_Base.php
Comment (spacedmonkey) [inline includes/Embed_Base.php]:
```suggestion
		if ( is_readable( WEBSTORIES_PLUGIN_DIR_PATH . 'includes/assets/embed.css' ) ) {
```
Massive nit pic, but it is best to use `is_readable`, as just because the file exists, doesn't mean php can read it. :P

**Comment by spacedmonkey** [comment]
@swissspidy What about this -https://github.com/google/web-stories-wp/blob/182c8f77f6293609e07aea9097b55375b8053fa9/includes/Admin.php#L113-L133

This kind of change was the point of this change - https://github.com/google/web-stories-wp/pull/3879

Does this need to change as well, so the preview matches the end result 

https://github.com/google/web-stories-wp/blob/182c8f77f6293609e07aea9097b55375b8053fa9/assets/src/story-embed-block/block/storyPlayer.js#L27-L47

**Comment by swissspidy** [comment]
We can't use the same markup in the block editor and on the frontend 1:1.

The idea was / would be along these lines

* Block `Edit` component: allows embedding and previewing a story, with edit controls etc.
* Block `Save` function: prints simple  markup as a fallback so that there's still _some_ content when the user deactivates the plugin  
  _Aside: I do realize now, while writing this, that we should probably just print an image or link here, not `<amp-story-player>`. Or even `return null`. I'll take a closer look._
* Frontend: Dynamic render callback to enhance the simple markup

The difference in markup between Frontend & `Save` is why we can't do #3879

However, I see now that the responsive part of this PR also needs to be reflected in the `Edit` component, otherwise the block is too large on smaller screens in the block editor:

Screenshot 2020-10-16 at 12 34 08

**Comment by spacedmonkey** [comment]
But if we save `<amp-story-player>` into content and deactivate the plugin, would the player even render without the script / style tags this plugin adds. Don't you think that is would safe just to save the image in content, which always works? Meaning #3879 can be used, by updated to use the `Story_rendered\Image` class. Only having an image in content, also means we don't have to worry about WordPress stripping out the amp-story-player tags on save. 

I think that the preview in the blocker editor, I think this is required. The preview should be as much like the output as possible and should be responsive if gutenberg is used on a mobile.

**Comment by swissspidy** [comment]
> But if we save `<amp-story-player>` into content and deactivate the plugin, would the player even render without the script / style tags this plugin adds. Don't you think that is would safe just to save the image in content, which always works?

Yeah that's what I said/meant in my comment:

> _Aside: I do realize now, while writing this, that we should probably just print an image or link here, not `<amp-story-player>`. Or even `return null`. I'll take a closer look._

**Comment by spacedmonkey** [comment]
> Yeah that's what I said/meant in my comment:

Sorry, I was agreeing with you and expanding, asking if we should use an image / `Story_rendered\Image` class.

**Comment by swissspidy** [comment]
If we use the `Image` class I'd be happy I think :-) I'll resurrect that ticket then.

File: includes/Block/Embed_Block.php
Comment (spacedmonkey) [inline includes/Block/Embed_Block.php:59]:
```suggestion
		$this->register_style( self::SCRIPT_HANDLE, [ self::STORY_PLAYER_HANDLE, self::SCRIPT_HANDLE ] );
```
No need to use the base class, the sub class self works just as well.

Comment (swissspidy) [reply includes/Block/Embed_Block.php:59]: Those are not the same.

`Embed_Block::SCRIPT_HANDLE` is `web-stories-embed-block` - for the block JS in Gutenberg

`Embed_Base::SCRIPT_HANDLE` is `web-stories-embed` - for the responsive CSS (and eventually JS) for embeds (used on frontend and backend). it's called `script_handle` because we might want to use it for jS too.

Comment (spacedmonkey) [reply includes/Block/Embed_Block.php:59]: Why was `SCRIPT_HANDLE` const reused here?

Comment (swissspidy) [reply includes/Block/Embed_Block.php:59]: This is one of the drawbacks of `Embed_Base` doing two things...

I explained it above.

I can rename `Embed_Base::SCRIPT_HANDLE` to `Embed_Base::STYLE_HANDLE` but it's still not ideal

Comment (spacedmonkey) [reply includes/Block/Embed_Block.php:59]: Doesn't bother me that much.

**Review by spacedmonkey — APPROVED** [review]
👍

**Pull request merged**
